unittests/Interpreter/InterpreterInfra.h

```diff
@@ -0,0 +1,153 @@
+      llvm::KnownBits ValueKB;
+
+      EvalValueKB(llvm::KnownBits Val);
+      EvalValueKB(EvalValue &Val);
+      llvm::KnownBits getValueKB();
+    };
+
+    class KBTesting {
+      // wrapper over @llvm::KnownBitsAnalysis::mergeKnownBits()
+      EvalValueKB merge(EvalValueKB a, EvalValueKB b);
+
+      llvm::KnownBits setLowest(llvm::KnownBits x);
+      llvm::KnownBits clearLowest(llvm::KnownBits x);
+
+      EvalValueKB bruteForce(llvm::KnownBits x, llvm::KnownBits y, Inst::Kind Pred);
+    public:
+      static bool nextKB(llvm::KnownBits &x);
+      bool testFn(Inst::Kind pred);
+    };
+
+
+
+    class CRTesting {
+      bool ok(const llvm::ConstantRange &R, const bool Table[]);
+
+      // Find the largest hole and build a llvm::ConstantRange around it
+      llvm::ConstantRange bestCR(const bool Table[], const int Width);
+      llvm::ConstantRange exhaustive(const llvm::ConstantRange &L, const llvm::ConstantRange &R,
+			       Inst::Kind pred, const llvm::ConstantRange &Untrusted);
+
+      void check(const llvm::ConstantRange &L, const llvm::ConstantRange &R, Inst::Kind pred,
+		 double &FastBits, double &PreciseBits, int &Count, int &PreciseCount);
+
+    public:
+
+      static llvm::ConstantRange nextCR(const llvm::ConstantRange &CR);
+      bool testFn(Inst::Kind pred);
+    };
+
+ namespace TestingUtil {
+
+    void exhaustiveKBCRReduction(llvm::KnownBits &KB, llvm::ConstantRange &CR);
+
+    template<int SetSize>
+    std::bitset<SetSize> concretizeKB(const llvm::KnownBits &KB) {
+      assert(SetSize == (1 << KB.getBitWidth()));
+
+      std::bitset<SetSize> ResultSet(1);
+      for (int Cnt = 0; Cnt < SetSize; Cnt++) {
+
+	llvm::KnownBits ValueKB;
+	auto Tmp = llvm::APInt(KB.getBitWidth(), Cnt);
+	ValueKB.One = Tmp;
+	ValueKB.Zero = ~Tmp;
+
+	if (KnownBitsAnalysis::isConflictingKB(ValueKB, KB))
+	  ResultSet[Cnt] = 0; // exclude this guy
+      }
+
+      return ResultSet;
+    }
+
+    template<int SetSize>
+    std::bitset<SetSize> concretizeCR(const llvm::ConstantRange &CR) {
+      assert(SetSize == (1 << CR.getBitWidth()));
+
+      std::bitset<SetSize> ResultSet(0);
+      if (CR.isFullSet()) {
+	ResultSet.set();
+	return ResultSet;
+      }
+
+      for (int Cnt = 0; Cnt < SetSize; Cnt++) {
+	if (CR.contains(llvm::APInt(WIDTH, Cnt)))
+	  ResultSet[Cnt] = 1; // include this guy
+      }
+      return ResultSet;
+    }
+
+    template<int SetSize>
+    llvm::ConstantRange abstractizeCR(std::bitset<SetSize> &ResultSet) {
+      llvm::ConstantRange ResultCR(WIDTH, false);
+      for (unsigned I = 0; I < SetSize; I++) {
+	if (ResultSet[I]) {
+	  ResultCR = ResultCR.unionWith(llvm::APInt(WIDTH, I));
+	}
+      }
+
+      return ResultCR;
+    }
+
+    template<int SetSize>
+    llvm::KnownBits abstractizeKB(std::bitset<SetSize> &ResultSet) {
+      llvm::KnownBits ResultKB(WIDTH);
+      if (ResultSet.none())
+	return ResultKB;
+
+      unsigned I = 0;
+      while (!ResultSet[I++]);
```

Comment:
I don't understand this loop, or the use of its results

Reply:
ResultSet is a bitvector. It's bit is 1 if the element is present and 0 otherwise. Here, we are finding the first element which is present and then initializing the value of {And,Or}Result later on with this index.

Reply:
but why? why is this the right thing to do?

Reply:
It's the right thing to do because we want first non-zero bit to initialize And/Or Result with. Eg: If WIDTH=4, the result set will be 16 bits long. Let's say that only number 4 and 12 are present in the ResultSet, and we want to abstractize it. In this case, the ResultSet will be : `0000100000001000`. We want to get the AndResult and OrResult of 4 and 12 which means we quickly want to access the first non-zero bit so that we can initialize `AndResult` and `OrResult` with "some" number and then keep ANDing or ORing as we encounter more non-zero bits in the ResultSet.

Reply:
this is just a weird way to do it Pranav, you can make this a lot simpler:
- eliminate the first loop
- in the second loop, just and together the ones (as you are doing) and then also and together the zeros, and don't flip the bits at the end

please do this.

Reply:
the first loop is not just useless but also confusing

Reply:
okay. i have simplified this. see the updated PR